distributed/tests/test_worker_state_machine.py

```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
     ]
 
 
+def test_constrained_task_handles_resources_on_success(ws_with_running_task):
+    ws = ws_with_running_task
+    assert ws.available_resources == {"R": 0}
+    # Ensure that x transitions through the ``constrained`` state
+    assert_story(
+        ws.story("x"),
+        [
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "waiting", {"x": "constrained"}),
+            ("x", "waiting", "constrained", "constrained", {"x": "executing"}),
+            ("x", "constrained", "executing", "executing", {}),
+        ],
+        strict=False,
+    )
+
+    ws.handle_stimulus(
+        ExecuteSuccessEvent(
+            key="x",
+            value=None,
+            start=0.0,
+            stop=1.0,
+            nbytes=8,
+            type=None,
+            stimulus_id="success",
+        ),
+    )
+    assert ws.tasks["x"].state == "memory"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.xfail(reason="distributed#6565")
+def test_constrained_task_handles_resources_on_failure(ws_with_running_task):
+    ws = ws_with_running_task
+    assert ws.available_resources == {"R": 0}
+    # Ensure that x transitions through the ``constrained`` state
+    assert_story(
+        ws.story("x"),
+        [
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "waiting", {"x": "constrained"}),
+            ("x", "waiting", "constrained", "constrained", {"x": "executing"}),
+            ("x", "constrained", "executing", "executing", {}),
+        ],
+        strict=False,
+    )
+
+    ws.handle_stimulus(
+        ExecuteFailureEvent(
+            key="x",
+            start=0.0,
+            stop=1.0,
+            exception=Serialize(ValueError("x")),
+            traceback=Serialize("Now, release resources"),
+            exception_text="exc text",
+            traceback_text="tb text",
+            stimulus_id="failure",
+        ),
+    )
+    assert ws.tasks["x"].state == "error"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.parametrize("state", ["executing", "long-running"])
+def test_resource_restricted_task_with_dependencies(ws, state):
+    ws.available_resources = {"R": 1}
+    ws.total_resources = {"R": 1}
+
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            key="y",
+            who_has={"x": ["127.0.0.1:1235"]},
+            nbytes={"x": 8},
+            resource_restrictions={"R": 1},
+            stimulus_id="compute",
+        )
+    )
+    assert ws.tasks["x"].state == "flight"
+    assert ws.tasks["y"].state == "waiting"
+    assert ws.available_resources == {"R": 1}
+
+    ws.handle_stimulus(
+        GatherDepSuccessEvent("gather-dep-done", "127.0.0.1:1235", 8, {"x": 1.0})
+    )
```

Comment:
This test has little benefit beyond this line. You should just test that this event return a Execute instruction, that the task state is 'executing', and that available_resources have been decreased.

Reply:
Fair point.